### PR TITLE
feat: add stringResolver

### PR DESCRIPTION
**Pull request by alexander-lozovsky**
Add `stringResolver` to control the rendering of the plain text, to handle cases when you need to preprocess the text before rendering (translation, sanitization, etc)

### CODE REVIEW

File: lib/src/types.ts [hunk 1]
```diff
@@ -39,6 +39,7 @@ export type Schema = {
 export type Options = {
   schema?: Schema;
   resolver?: (blok: SbBlok) => ComponentNode;
+  stringResolver?: (str: string) => ComponentNode;
```
Comment (edvinasjurele) [inline lib/src/types.ts:42]: I would rather call this function a `contentResolver` as it wraps content with additional logic.

File: lib/RichTextRenderer.astro [hunk 2]
```diff
@@ -19,11 +19,16 @@ export type Props = {
    ```
    */
   resolver: Options["resolver"];
+  stringResolver: Options["stringResolver"];
```
Comment (edvinasjurele) [inline lib/RichTextRenderer.astro:22]: Lets give it a thorough description, similar to `resolver` one.

File: lib/src/utils/resolveRichTextToNodes.ts [hunk 3]
```diff
@@ -109,7 +109,15 @@ export const resolveNode = (
...
     if (marks) {
-      let marked: ComponentNode[] = [{ content: text }];
+      let marked: ComponentNode[] = stringResolver
+        ? [
+            {
+              content: text,
+              ...stringResolver(text),
+            },
+          ]
+        : [{ content: text }];
```
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.ts:119]: both branch sides has same base of `[{ content: text }]` while the `stringResolver` usage is conditional to the left side, so maybe this could be simplified to the:

```js
let marked: ComponentNode[] = [
  {
    content: text,
    ...stringResolver?.(text),
  },
];
```

did not try the code, but in case it does not work it might need to additionally add conditional to not spread undefined: `...(stringResolver?.('d') || {}`

**Comment by edvinasjurele** [comment]
Also do not forget DCO, thus all commits has to be signed-off. Read the details and follow instructions.

File: demo/src/storyblok/RichText.astro [hunk 4]
```diff
@@ -52,5 +52,10 @@ const { text } = blok;
       props: { blok },
     };
   }}
+  stringResolver={() => {
+    return {
+      component: "span",
+    };
+  }}
```
Comment (edvinasjurele) [inline demo/src/storyblok/RichText.astro:59]: Just an optional idea. we might update the example with the full usage, meaning to take that string from fn prop and return an actual content value in the object.
I would also leave some small comment (since this is example) in this function like:

```js
 stringResolver={(content) => {
    // custom logic to manipulate content string value

    return {
      component: "span",
      content,
    };
  }}
```

or alternatively this could be part of JSDoc comment to showcase the example of possible application of this function.

File: lib/src/utils/resolveRichTextToNodes.spec.ts [hunk 5]
```diff
@@ -83,6 +83,58 @@ describe("resolveNode", () => {
...
+    // with marks and stringResolver
+    expect(
+      resolveNode(
+        {
+          text: "I am text",
+          type: "text",
+          marks: [{ type: "bold" }],
+        },
+        {
+          stringResolver: () => ({
```
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.spec.ts:116]: I see in the implementation that the function supports a text prop access: `stringResolver: (text) => ({` so it might be good idea to add one more test case for affecting that. similar to what the react alternative does https://www.npmjs.com/package/storyblok-rich-text-react-renderer#text-resolver

File: lib/src/types.ts (hunk 1 above)
Comment (edvinasjurele) [reply lib/src/types.ts:42]: After all, we might call the same as react alternative has https://www.npmjs.com/package/storyblok-rich-text-react-renderer#text-resolver which is `textResolver` as it indeed takes the `text` node wrapping:
```
textResolver?: (text?: string) => ComponentNode;
```

Comment (edvinasjurele) [reply lib/src/types.ts:42]: After all, we might call the same as react alternative has https://www.npmjs.com/package/storyblok-rich-text-react-renderer#text-resolver which is `textResolver` as it indeed takes the `text` node wrapping:
```
textResolver?: (text?: string) => ComponentNode;
```

**Comment by edvinasjurele** [comment]
Lets also update README with the information about the usage, and maybe some use cases.

Optionally, we should extend DEMO app with the example of such resolver too. We can do some simple text replacement example or similar simple case just to illustrate that it is possible to hook into text node and affect it.

File: lib/src/types.ts (hunk 1 above)
Comment (alexander-lozovsky) [reply lib/src/types.ts]: renamed it to textResolver

File: lib/RichTextRenderer.astro (hunk 2 above)
Comment (alexander-lozovsky) [reply lib/RichTextRenderer.astro]: done

File: demo/src/storyblok/RichText.astro (hunk 4 above)
Comment (alexander-lozovsky) [reply demo/src/storyblok/RichText.astro:60]: done

File: lib/src/utils/resolveRichTextToNodes.spec.ts (hunk 5 above)
Comment (alexander-lozovsky) [reply lib/src/utils/resolveRichTextToNodes.spec.ts]: done

File: lib/src/types.ts (hunk 1 above)
Comment (alexander-lozovsky) [reply lib/src/types.ts]: alright, I remembered why I named it `stringResolver`: in the react library there are two functions `defaultStringResolver` and `textResolver`. `textResolver` is to resolve **text nodes**, while `defaultStringResolver` is to resolve plain strings https://github.com/claus/storyblok-rich-text-react-renderer/blob/master/src/index.js#L104C14-L104C14. Documentation there is a bit misleading

File: lib/src/utils/resolveRichTextToNodes.ts (hunk 3 above)
Comment (alexander-lozovsky) [reply lib/src/utils/resolveRichTextToNodes.ts]: correct, simplified the implementation

Additionally, handled a corner case when both schema text node resolver and `textResolver` are provided, to not override each other. See the test cases

File: lib/src/types.ts (hunk 1 above)
Comment (edvinasjurele) [reply lib/src/types.ts]: Sure, its all fine, this API does not necessarily need to match 1:1 with Claus one. I am totally fine with having `textResolver` for, well, resolving text's (which basically the same as `defaultStringResolver` in react), and leaving existing availability for users to affect nodes via `schema.nodes.text` function.

File: lib/src/utils/resolveRichTextToNodes.spec.ts [hunk 6]
```diff
@@ -113,6 +113,83 @@ describe("resolveNode", () => {
...
+              component: "p",
+              props: { class: "class-1" },
+            }),
+          },
+        },
+        textResolver: (text) => ({
+          content: text.replace("{name}", "World"),
+          component: "span",
+          props: { class: "class-2" },
+        }),
```
Comment (edvinasjurele) [inline lib/src/utils/resolveRichTextToNodes.spec.ts:176]: Hm, this does look weird honestly, and I am sure it clashes with the existing API. The problem is that conceptually the `textResolver` should be meant for text resolving only, and should not return `ComponentNode` at all. This function is not about the node logic at all, but rather about manipulating the `content` property in whatever the node is.

to get same result you have described with solution we should better have the following API:
```
     schema: {
        nodes: {
          text: () => ({
            component: "p",
            props: { class: "class-1" },
          }),
        },
      },
      textResolver: (text) => text.replace("{name}", "World")
```
(the textResolver extends the capability of the node, but does not override it)

Alternatively, we can expect the shape of `textResolver?: (str: string) => Pick<ComponentNode, 'content'>;`
and the usage:
```
 textResolver: (text) => ({
    content: text.replace("{name}", "World"),
 }),
```
but again, this may misslead users that it is not the whole component node but rather just a content, so might be bad alternative.

What do you think?

Comment (alexander-lozovsky) [reply lib/src/utils/resolveRichTextToNodes.spec.ts:176]: it should return `ComponentNode` for cases when you example you want to support ICU syntax. So in this case all the texts need to be rendered inside some `.astro` component

I can suggest to support both outputs:
`textResolver?: (str: string) => ComponentNode | string;`
